File: experimental/ip-geofeed/internal/html_template/html.go
package html_template

import (
	"encoding/json"
	"fmt"
	"html/template"
	"os"
	"path/filepath"

	"ip-geofeed/internal/geofeed_validation"
)

// GenerateHTMLReport generates an HTML validation report from entries
func GenerateHTMLReport(entries []geofeed_validation.Entry, comments map[int]string, metadata geofeed_validation.Metadata, path string) error {
	// Parse the template file
	tmpl, err := template.ParseFiles("internal/html_template/report.html")
	if err != nil {
		return fmt.Errorf("parsing template file: %w", err)
	}
	commentsJson, err := json.Marshal(comments)
	if err != nil {
		return fmt.Errorf("marshaling comments to JSON: %w", err)
	}

	// Create the data structure to pass to the template
	data := struct {
		Entries  []geofeed_validation.Entry
		Comments template.JS
		Metadata geofeed_validation.Metadata
	}{
		Entries:  entries,
		Comments: template.JS(commentsJson),
		Metadata: metadata,
	}

	// Extract directory
	filePath := filepath.Join("run", "output", path)
	dirPath := filepath.Dir(filePath)

	// Create directories if they don't exist
	err = os.MkdirAll(dirPath, 0755)
	if err != nil {
		return fmt.Errorf("creating output directory: %w", err)
	}

	outputFile, err := os.Create(filePath)
	if err != nil {
		return fmt.Errorf("creating output file: %w", err)
	}
	defer outputFile.Close()

	// Execute the template with the data
	err = tmpl.Execute(outputFile, data)
	if err != nil {
		return fmt.Errorf("executing template: %w", err)
	}

	return nil
}

func GenerateNetnameHTMLTable(records geofeed_validation.Netname, path string) error {
	// Parse the template file
	tmpl, err := template.ParseFiles("internal/html_template/netname_table.html")
	if err != nil {
		return fmt.Errorf("parsing template file: %w", err)
	}

	// Create the data structure to pass to the template
	data := struct {
		Records []geofeed_validation.Record
	}{
		Records: records.Records,
	}

	// Extract directory
	filePath := filepath.Join("run", "output", path)
	dirPath := filepath.Dir(filePath)

	// Create directories if they don't exist
	err = os.MkdirAll(dirPath, 0755)
	if err != nil {
		return fmt.Errorf("creating output directory: %w", err)
	}

	outputFile, err := os.Create(filePath)
	if err != nil {
		return fmt.Errorf("creating output file: %w", err)
	}
	defer outputFile.Close()

	// Execute the template with the data
	err = tmpl.Execute(outputFile, data)
	if err != nil {
		return fmt.Errorf("executing template: %w", err)
	}

	return nil
}

File: experimental/ip-geofeed2/internal/html_template/html.py
import json
import os

from jinja2 import Environment, FileSystemLoader

from internal.geofeed_validation.validator import Entry, Metadata, Netname, ValidatorRecord


_TEMPLATE_DIR = os.path.join("internal", "html_template", "templates")
_env = Environment(loader=FileSystemLoader(_TEMPLATE_DIR), autoescape=True)


def generate_html_report(
    entries: list[Entry],
    comments: dict[int, str],
    metadata: Metadata,
    path: str,
) -> None:
    """Generates an HTML validation report from entries."""
    tmpl = _env.get_template("report.html")

    comments_json = json.dumps(comments)

    html = tmpl.render(
        entries=entries,
        comments_json=comments_json,
        metadata=metadata,
    )

    file_path = os.path.join("run", "output", path)
    os.makedirs(os.path.dirname(file_path), exist_ok=True)

    with open(file_path, "w") as f:
        f.write(html)


def generate_netname_html_table(records: list[ValidatorRecord], path: str) -> None:
    """Generates an HTML table for netname records."""
    tmpl = _env.get_template("netname_table.html")

    html = tmpl.render(records=records)

    file_path = os.path.join("run", "output", path)
    os.makedirs(os.path.dirname(file_path), exist_ok=True)

    with open(file_path, "w") as f:
        f.write(html)


def generate_source_html_table(netnames: list[Netname], source: str, path: str) -> None:
    """Generates an HTML table for source/RIR records."""
    tmpl = _env.get_template("source_table.html")

    html = tmpl.render(netnames=netnames, source=source)

    file_path = os.path.join("run", "output", path)
    os.makedirs(os.path.dirname(file_path), exist_ok=True)

    with open(file_path, "w") as f:
        f.write(html)
